src/Database/LSMTree.hs

```diff
@@ -0,0 +1,551 @@
+  where
+    checkBlobRefType _ (BlobRef (ref :: Internal.WeakBlobRef m h'))
+      | Just Refl <- eqT @h @h' = ref
+    checkBlobRefType i _ = throw (Internal.ErrBlobRefInvalid i)
+
+{-------------------------------------------------------------------------------
+  Snapshots
+-------------------------------------------------------------------------------}
+
+{-# SPECIALISE createSnapshot ::
+     (Common.Labellable (k, v, b), ResolveValue v)
+  => SnapshotName
+  -> Table IO k v b
+  -> IO () #-}
+createSnapshot :: forall m k v b.
+     ( IOLike m
+     , Common.Labellable (k, v, b)
+     , ResolveValue v
+     )
+  => SnapshotName
+  -> Table m k v b
+  -> m ()
+createSnapshot snap (Internal.Table' t) =
+    void $ Internal.createSnapshot (resolve (Proxy @v)) snap label Internal.SnapFullTable t
+  where
+    label = Internal.SnapshotLabel $ Common.makeSnapshotLabel (Proxy @(k, v, b))
+
+{-# SPECIALISE openSnapshot ::
+     (Common.Labellable (k, v, b), ResolveValue v)
+  => Session IO
+  -> Common.TableConfigOverride
+  -> SnapshotName
+  -> IO (Table IO k v b ) #-}
+openSnapshot :: forall m k v b.
+     ( IOLike m
+     , Common.Labellable (k, v, b)
+     , ResolveValue v
+     )
+  => Session m
+  -> Common.TableConfigOverride -- ^ Optional config override
+  -> SnapshotName
+  -> m (Table m k v b)
+openSnapshot (Internal.Session' sesh) override snap =
+    Internal.Table' <$!> Internal.openSnapshot sesh label Internal.SnapFullTable override snap (resolve (Proxy @v))
+  where
+    label = Internal.SnapshotLabel $ Common.makeSnapshotLabel (Proxy @(k, v, b))
+
+{-------------------------------------------------------------------------------
+  Mutiple writable tables
+-------------------------------------------------------------------------------}
+
+{-# SPECIALISE duplicate ::
+     Table IO k v b
+  -> IO (Table IO k v b) #-}
+duplicate ::
+     IOLike m
+  => Table m k v b
+  -> m (Table m k v b)
+duplicate (Internal.Table' t) = Internal.Table' <$!> Internal.duplicate t
+
+{-------------------------------------------------------------------------------
+  Table union
+-------------------------------------------------------------------------------}
+
+{-# SPECIALISE union ::
+     ResolveValue v
+  => Table IO k v b
+  -> Table IO k v b
+  -> IO (Table IO k v b) #-}
+union :: forall m k v b.
+     ( IOLike m
+     , ResolveValue v
+     )
+  => Table m k v b
+  -> Table m k v b
+  -> m (Table m k v b)
+union = error "union: not yet implemented" $ union @m @k @v @b
+
+{-------------------------------------------------------------------------------
+  Monoidal value resolution
+-------------------------------------------------------------------------------}
+
+resolve ::  forall v. ResolveValue v => Proxy v -> Internal.ResolveSerialisedValue
+resolve = coerce . resolveValue
+
+-- | Newtype wrapper for use with deriving via, which resolves values as a
+-- 'Data.Semigroup.First' semigroup.
```

Comment:
Another way than DerivingVia would be to use `Table k (ResolveAsFirst v) b`, which we could also provide a type synonym for. Something like this?

```haskell
type SimpleTable k v = Table k (ResolveAsFirst v) Void
type BlobTable k v b = Table k (ResolveAsFirst v) b
type MonoidalTable k v = Table k v Void
```

But not too important for now, I think, as the main purpose of this module is for testing. Later on, if we really want this module to be the entry point for users, we can still work on making that a nicer experience.